### ISSUE TITLE
关于交叉熵

### ISSUE THREAD

**Issue by linhaojia13** [yangdechuan/fast-gradient-sign-method-tensorflow#1]
作者你好，关于代码中对交叉熵的使用我有一点疑问，代码的72行把所有图片的交叉熵都加总到一起：
        cross_entropy = tf.reduce_mean(-tf.reduce_sum(y_ * tf.log(y), axis=[1]))
在95行则是利用这个加总的交叉熵来计算每一个图片的对抗扰动：
        perb = self.epsilon * tf.sign(tf.gradients(cross_entropy, x)[0])  # calculate perturbation
但事实上每张图片之间是相互独立的，在某一张图片上施加扰动并不会影响另一张图片的loss，所以我认为72行这里应该改成：
        cross_entropy = -tf.reduce_sum(y_ * tf.log(y), axis=[1])
保留第一个维度以保证每张图片的loss回传的梯度只影响到其自身。
请问你怎么看？

**Comment by yangdechuan** [comment]
1. cross_entropy = -tf.reduce_sum(y_ * tf.log(y), axis=[1])返回的是一个向量，不能作为loss吧；
2. 关于“在某一张图片上施加扰动并不会影响另一张图片的loss”，这样扰动是可以增加整个batch数据的loss的；
3. 看了下原文好像是每次对一个样本进行扰动，那么扰动的时候可以将batch_size设置为1。

**Comment by webbrain-one** [comment]
感谢指出！你说得对，原来用整批平均交叉熵会让梯度按 batch 平均，可能把不同样本的梯度混在一起。已开 PR #2，修复方法是保留每张图片的 loss（去掉 `reduce_mean`），这样每个样本的 FGSM 扰动只由自身 loss 的梯度决定。
